src/Microsoft.Health.Fhir.Shared.Api/Features/Security/SecurityProvider.cs

```diff
@@ -122,75 +126,77 @@ private void AddOAuthSecurityService(ListedCapabilityStatement statement)
-                {
-                    url = Constants.SmartOAuthUriExtension,
-                    extension = new[]
-                    {
-                        new
-                        {
-                            url = Constants.SmartOAuthUriExtensionToken,
-                            valueUri = tokenEndpoint,
-                        },
-                        new
-                        {
-                            url = Constants.SmartOAuthUriExtensionAuthorize,
-                            valueUri = authorizationEndpoint,
-                        },
-                    },
-                };
+                // SMART on FHIR is not configured.
+                // Set the restful-security-service code to OAuth.
+                codableConceptInfo.Coding.Add(
+                    _modelInfoProvider.Version == FhirSpecification.Stu3
+                    ? Constants.RestfulSecurityServiceStu3OAuth
+                    : Constants.RestfulSecurityServiceOAuth);
+
+                // Fallback to OpenID configuration.
+                OpenIdConfigurationResponse openIdConfiguration = _wellKnownConfigurationProvider.GetOpenIdConfigurationAsync(CancellationToken.None).GetAwaiter().GetResult();
+                authorizationEndpoint = openIdConfiguration?.AuthorizationEndpoint;
+                tokenEndpoint = openIdConfiguration?.TokenEndpoint;
+            }
 
-                security.Extension.Add(JObject.FromObject(smartExtension));
+            if (authorizationEndpoint != null && tokenEndpoint != null)
+            {
+                var extension = new JArray();
+                AddExtension(extension, Constants.SmartOAuthUriExtensionToken, tokenEndpoint);
+                AddExtension(extension, Constants.SmartOAuthUriExtensionAuthorize, authorizationEndpoint);
+                AddExtension(extension, Constants.SmartOAuthUriExtensionRegister, registrationEndpoint);
+                AddExtension(extension, Constants.SmartOAuthUriExtensionManage, managementEndpoint);
+                AddExtension(extension, Constants.SmartOAuthUriExtensionRevoke, revocationEndpoint);
+                AddExtension(extension, Constants.SmartOAuthUriExtensionIntrospect, introspectionEndpoint);
+
+                security.Extension.Add(JObject.FromObject(new { url = Constants.SmartOAuthUriExtension, extension }));
             }
             else
             {
-                _logger.LogWarning("The OpenId Configuration request from \"{OpenIdConfigurationUrl}\" returned an {StatusCode} status code.", openIdConfigurationUrl, openIdConfigurationResponse.StatusCode);
-                openIdConfigurationResponse.Dispose();
                 throw new OpenIdConfigurationException();
             }
 
             restComponent.Security = security;
         }
+
+        private static void AddExtension(JArray extension, string key, Uri value)
+        {
+            if (!string.IsNullOrWhiteSpace(key) && value != null)
```

Comment:
is the key check on constant keys needed?

Reply:
I code very defensively. Because this method does not depend on the constants itself, it is a good practice to assume values can be null.